src/patient-registration/patient-registration.component.tsx

```diff
@@ -0,0 +1,138 @@
+import { Address } from './field/address/address.component';
+import styles from './patient-registration.css';
+
+const IDENTIFIER_TYPE: string = '05a29f94-c0ed-11e2-94be-8c13b969e334';
+
+interface PatientRegistrationProps {}
+
+export function PatientRegistration(props: PatientRegistrationProps) {
+  const [unknown, setUnknown] = useState<boolean>(false);
+  const [identifier, setIdentifier] = useState<Patient['identifiers'][0]['identifier']>('');
+  const [location, setLocation] = useState<Patient['identifiers'][0]['location']>('');
+  const [name, setName] = useState<Patient['person']['names'][0]>({
+    preferred: true,
+    givenName: '',
+    middleName: '',
+    familyName: '',
+  });
+  const [additionalName, setAdditionalName] = useState<Patient['person']['names'][1]>({
+    preferred: false,
+    givenName: '',
+    middleName: '',
+    familyName: '',
+  });
+  const [gender, setGender] = useState<Patient['person']['gender']>('');
+  const [birthdate, setBirthdate] = useState<Patient['person']['birthdate']>(null);
+  const [birthdateEstimated, setBirthdateEstimated] = useState<Patient['person']['birthdateEstimated']>(false);
+  const [birthtime, setBirthtime] = useState<Patient['person']['birthtime']>(null);
+  const [address, setAddress] = useState<Patient['person']['addresses'][0]>({
+    address1: '',
+    address2: '',
+    cityVillage: '',
+    stateProvince: '',
+    postalCode: '',
+    country: '',
+    latitude: '',
+    longitude: '',
+    startDate: null,
+    endDate: null,
+  });
+
+  useEffect(() => {
+    const abortController = new AbortController();
+    getCurrentUserLocation(abortController).then(
+      ({ data }) => setLocation(data.sessionLocation.uuid),
+      createErrorHandler(),
+    );
+    return () => abortController.abort();
+  }, []);
+
+  useEffect(() => {
+    const abortController = new AbortController();
+    getUniquePatientIdentifier(abortController).then(
+      ({ data }) => setIdentifier(data.identifiers[0]),
+      createErrorHandler(),
+    );
+    return () => abortController.abort();
+  }, []);
+
+  const onFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    const abortController = new AbortController();
+    const patient: Patient = {
+      identifiers: [
+        {
+          identifier: identifier,
+          identifierType: IDENTIFIER_TYPE,
+          location: location,
+        },
+      ],
+      person: {
+        names: [name, additionalName],
+        gender: gender,
+        birthdate: birthdate,
+        birthdateEstimated: birthdateEstimated,
+        birthtime: birthtime,
+        addresses: [address],
+      },
+    };
+
+    savePatient(abortController, patient).then(
+      response => response.status == 201 && navigate(response.data.uuid),
+      createErrorHandler(),
+    );
+  };
+
+  const navigate = (patientUuid: string) => {
+    window.location.href = `/openmrs/spa/patient/${patientUuid}/chart`;
```

Comment:
I think this leads to another SPA part?

I would rather use `useHistory()` and `history.push`.

Maybe you tried and did not work; this is because I think we have not wrapped this thing in a react router. All you'd need is to use a `BrowserRouter` as a top-level component.

We could also do that in a subsequent PR.

**Remark**: Just to be clear why this is a bad idea: right now we have a SPA, where S is for Single. There is a huge startup cost, which we'll only benefit from if we have much reduced cost when changing "pages". Now setting `window.location.href` will trigger a full page load, as in an MPA (M = Multi) scenario. We would lose all the page load effort for fully loading a page again which was made for SPA.

Reply:
We've just implemented the link by wrapping the `<Root /> `in `<BrowserRouter />` and then passing `useHistory()` as props to the `PatientRegistration` component. The component then calls `history.push()` when submitting the form that takes the user to the `PatientChart `component for the patient they just registered.

Reply:
As `useHistory` uses a context there is no need to pass that down. Just call `useHistory` from the component that needs to call `history.push`.